compositor_pipeline/src/pipeline/output/mp4.rs

```diff
@@ -0,0 +1,213 @@
+use std::path::PathBuf;
+
+use compositor_render::OutputId;
+use crossbeam_channel::Receiver;
+use ffmpeg_next::Packet;
+use log::error;
+
+use crate::{
+    error::OutputInitError,
+    pipeline::{AudioCodec, EncodedChunk, EncodedChunkKind, EncoderOutputEvent, VideoCodec},
+};
+
+#[derive(Debug, Clone)]
+pub struct Mp4OutputOptions {
+    pub output_path: PathBuf,
+    pub video: Option<Mp4VideoTrack>,
+    pub audio: Option<Mp4AudioTrack>,
+}
+
+#[derive(Debug, Clone)]
+pub struct Mp4VideoTrack {
+    pub codec: VideoCodec,
+    pub width: u32,
+    pub height: u32,
+}
+
+#[derive(Debug, Clone)]
+pub struct Mp4AudioTrack {
+    pub codec: AudioCodec,
+}
+
+pub enum Mp4OutputVideoTrack {
+    H264 { width: u32, height: u32 },
+}
+
+pub struct Mp4WriterOptions {
+    pub output_path: PathBuf,
+    pub video: Option<Mp4OutputVideoTrack>,
+}
+
+pub struct Mp4FileWriter;
+
+impl Mp4FileWriter {
+    pub fn new(
+        output_id: &OutputId,
+        options: Mp4OutputOptions,
+        packets_receiver: Receiver<EncoderOutputEvent>,
+    ) -> Result<Self, OutputInitError> {
+        if options.output_path.exists() {
+            return Err(OutputInitError::Mp4PathExist {
+                path: options.output_path.to_string_lossy().into_owned(),
+            });
+        }
+
+        let (mut output_ctx, video_stream_id, audio_stream_id) = Self::init_ffmpeg_output(options)?;
+        let mut playing_streams = [video_stream_id, audio_stream_id]
+            .into_iter()
+            .filter(Option::is_some)
+            .count();
+
+        std::thread::Builder::new()
+            .name(format!("mp4 writer thread for output {}", output_id))
+            .spawn(move || {
+                for packet in packets_receiver {
+                    match packet {
+                        EncoderOutputEvent::Data(chunk) => {
+                            Self::write_chunk(
+                                chunk,
+                                video_stream_id,
+                                audio_stream_id,
+                                &mut output_ctx,
+                            );
+                        }
+                        EncoderOutputEvent::AudioEOS => playing_streams -= 1,
+                        EncoderOutputEvent::VideoEOS => playing_streams -= 1,
```

Comment:
Storing whether EOS was received via number seems wrong, if semantically this is bool it either was received or not. I would use `Option<bool>` per stream where None represents no stream, Some(true) received EOS, Some(false) not EOS yet